Pattern changes caused by commit: 185eca5d1fa7e384bb888c144d06abbced0fd577

From: Facade-1
To:   Facade-0

From: Flyweight-3
To:   Flyweight-4


=========================
       NEW GIT LOG
=========================

This commit refers to file: VALID-5781.txt 

commit 185eca5d1fa7e384bb888c144d06abbced0fd577
Author: Jonathan Ellis <jbellis@apache.org>

    prevent slow clients from postponing shutdown indefinitely
    patch by jbellis; reviewed by brandonwilliams for CASSANDRA-3727



==================================
 Issue CASSANDRA-3727 Description 
=======================================

Project: Cassandra
-----------------

-----------------
Title: [CASSANDRA-3727] Fix unit tests failure
-----------------

-----------------
Summary: Fix unit tests failure
-----------------

-----------------
Issue type: Bug
-----------------

-----------------
Current status: Resolved
-----------------

-----------------
Created at: Wed, 11 Jan 2012 09:39:03 +0000
-----------------

-----------------
Resolved at: Thu, 12 Jan 2012 09:25:04 +0000
-----------------

-----------------
Assigned to: Unassigned
-----------------

-----------------
Description: 

On current 1.0 branch (and on my machine: Linux), I have the following unit test
failures:
<ul>	<li>CliTest and EmbeddedCassandraTest: they both first kind of pass (JUnit
first prints a message with no failures in it), then hang until JUnit timeout and fails
with a 'Timeout occurred'. In other word, the tests themselves are passing, but something
they do prevents the process to exit cleanly leading to a JUnit timeout. I don't want to
discard that as not a problem, because if something can make the process not exit cleanly,
this can be a pain for restarts (and in particular upgrades) and hence would be basically
a regression. I'm marking the ticket as blocker (for the release of 1.0.7) mostly because
of this one.</li>	<li>SystemTableTest: throws an assertionError. I haven't checked yet, so
that could be an easy one to fix.</li>	<li>RemoveTest: it fails, saying that
'/127.0.0.1:7010 is in use by another process' (consistently). But I have no other process
running on port 7010. It's likely just of problem of the test, but it's new and in the
meantime removes are not tested.</li>	<li>I also see a bunch of stack trace with errors
like:<div class="preformatted panel" style="border-width: 1px;"><div
class="preformattedContent panelContent"><pre>    [junit] ERROR 10:01:59,007 Fatal
exception in thread Thread[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1-hc-1-Index.db to
/home/mcmanus/Git/cassandra/build/test/cassandra/data/Keyspace1/backups/Indexed1-hc-1-Index.db
(errno 17)</pre></div></div>
(with SSTableReaderTest). This does not make the tests fail,
but it is still worth investigating. It may be due to <a
href="https://issues.apache.org/jira/browse/CASSANDRA-3101" title="Should check for errors
when calling /bin/ln" class="issue-link"
data-issue-key="CASSANDRA-3101"><del>CASSANDRA-3101</del></a>.
</li></ul> 

-----------------

-----------------
Comments: 

New Comment: 
CliTest and others should be timeouting because of newly added shutdown hook. 


New Comment: 
Are you saying that they should be timeouting as in 'having them timeouting is a feature'
or are you just pointing out the likely source of the problem? In the latter, do  you
remember the ticket that introduced those (or better, have a fix for it)? 


New Comment: 
I'm saying that it's likely source of the problem and for CliTest fix would be pretty
straightforward, make CliMain to disconnect properly after tests are done (I'm going to
attach a patch fixing patch in a few minutes to this ticket), I don't know about other
tests tho. 


New Comment: 
fixed CliTest timeout by correctly closing transport connection which allows Cassandra
shutdown hook to proceed without waiting for RPC connections to close. 


New Comment: 
Hmm.  Can we fix the thrift shutdown instead to not wait for sockets to be closed nicely? 


New Comment: 
I guess we can, but I thought that it's kind of feature that it does wait... 


New Comment: 
Definitely not, otherwise there is no way to shut down if clients stay connected 


New Comment: 
(The thrift shutdown was introduced for <a
href="https://issues.apache.org/jira/browse/CASSANDRA-3335" title="ThreadPoolExecutor
creates threads as non-daemon and will block on shutdown by default" class="issue-link"
data-issue-key="CASSANDRA-3335"><del>CASSANDRA-3335</del></a>) 


New Comment: 
<blockquote>Can we fix the thrift shutdown instead to not wait for sockets to be closed
nicely</blockquote>I'm not sure why this isn't how it already works.  From
CustomTThreadPoolServer.WorkerProcess:<div class="code panel" style="border-width:
1px;"><div class="codeContent panelContent"><pre class="code-java">.               <span
class="code-keyword">while</span> (!stopped_ &amp;&amp; processor.process(inputProtocol,
outputProtocol))                {                    inputProtocol =
inputProtocolFactory_.getProtocol(inputTransport);                    outputProtocol =
outputProtocolFactory_.getProtocol(outputTransport);                }</pre></div></div>In
other words, as soon as stopped_ is set (by the stop() method), each thread should finish
the current request but not accept more. 


New Comment: 
It's easy to test - run cassandra and in other terminal session connect to it using CLI
and try Ctrl-C Cassandra server without closing CLI. 


New Comment: 
So, I was over-optimistic in <a
href="https://issues.apache.org/jira/browse/CASSANDRA-3335" title="ThreadPoolExecutor
creates threads as non-daemon and will block on shutdown by default" class="issue-link"
data-issue-key="CASSANDRA-3335"><del>CASSANDRA-3335</del></a> when I thought I could get
by without a MessagingService shutdown method.  The problem is that although my changes
there do work to prevent accepting new connections, and to stop work on existing
connections <b>after the first command post-shutdown</b>, that's not good enough in this
case since the client is just sitting on its connection and never sends another
command.So, this patch renames MS.waitForCallbacks() back to shutdown(), and refuses to
add new callbacks after that.  However, the analysis on 3335 that there's no good way to
deal with an exception here, so we do this instead in ExpiringMap:<div class="code panel"
style="border-width: 1px;"><div class="codeContent panelContent"><pre class="code-java">. 
 <span class="code-keyword">public</span> V put(K key, V value, <span
class="code-object">long</span> timeout)    {        <span class="code-keyword">if</span>
(shutdown)        {            <span class="code-comment">// StorageProxy isn<span
class="code-quote">'t equipped to deal with <span class="code-quote">"I'</span>m nominally
alive, but I can't send any messages out."</span></span>            <span
class="code-comment">// So we'll just sit on <span class="code-keyword">this</span> thread
<span class="code-keyword">for</span> until the <span class="code-keyword">rest</span> of
the server shutdown completes.</span>            <span class="code-comment">//</span>     
      <span class="code-comment">// See comments in CustomTThreadPoolServer.serve,
CASSANDRA-3335, and CASSANDRA-3727.</span>            <span
class="code-keyword">try</span>            {                <span
class="code-object">Thread</span>.sleep(<span class="code-object">Long</span>.MAX_VALUE); 
          }            <span class="code-keyword">catch</span> (InterruptedException e)   
        {                <span class="code-keyword">throw</span> <span
class="code-keyword">new</span> AssertionError(e);            }        }       
CacheableObject&lt;V&gt; previous = cache.put(key, <span class="code-keyword">new</span>
CacheableObject&lt;V&gt;(value, timeout));        <span class="code-keyword">return</span>
(previous == <span class="code-keyword">null</span>) ? <span
class="code-keyword">null</span> : previous.getValue();    }</pre></div></div>Then, we
switch the Thrift executor (and all DTPE instances) to use daemon threads, and remove the
wait-for-WorkerProcess threads code from CustomTThreadPoolServer.serve:<div class="code
panel" style="border-width: 1px;"><div class="codeContent panelContent"><pre
class="code-java">.       <span class="code-comment">// Thrift's <span
class="code-keyword">default</span> shutdown waits <span class="code-keyword">for</span>
the WorkerProcess threads to complete.  We <span class="code-keyword">do</span>
not,</span>        <span class="code-comment">// because doing that allows a client to
hold our shutdown <span class="code-quote">"hostage"</span> by simply not sending</span>  
     <span class="code-comment">// another message after stop is called (since process
will block indefinitely trying to read</span>        <span class="code-comment">// the
next meessage header).</span>        <span class="code-comment">//</span>        <span
class="code-comment">// The <span class="code-quote">"right"</span> fix would be to update
thrift to set a socket timeout on client connections</span>        <span
class="code-comment">// (and tolerate unintentional timeouts until stopped_ is set).  But
<span class="code-keyword">this</span> requires deep</span>        <span
class="code-comment">// changes to the code generator, so simply setting these threads to
daemon (in our custom</span>        <span class="code-comment">// CleaningThreadPool) and
ignoring them after shutdown is good enough.</span>        <span
class="code-comment">//</span>        <span class="code-comment">// Remember, our goal on
shutdown is not necessarily that each client request we receive</span>        <span
class="code-comment">// gets answered first [to <span class="code-keyword">do</span> that,
you should redirect clients to a different coordinator</span>        <span
class="code-comment">// first], but rather (1) to make sure that <span
class="code-keyword">for</span> each update we ack as successful, we generate</span>      
 <span class="code-comment">// hints <span class="code-keyword">for</span> any
non-responsive replicas, and (2) to make sure that we quickly stop</span>        <span
class="code-comment">// accepting client connections so shutdown can <span
class="code-keyword">continue</span>.  Not waiting <span class="code-keyword">for</span>
the WorkerProcess</span>        <span class="code-comment">// threads here accomplishes
(2); MessagingService's shutdown method takes care of (1).</span>        <span
class="code-comment">//</span>        <span class="code-comment">// See CASSANDRA-3335 and
CASSANDRA-3727.</span></pre></div></div>Finally, this patch also updates Memtable's
memorymeter thread to use the newly daemonized DTPE for good measure, since there's no
reason to ever block shutdown for that either. 


New Comment: 
+1 


New Comment: 
committed; leaving open for other test failures 


New Comment: 
SystemTableTest failure taken care of on <a
href="https://issues.apache.org/jira/browse/CASSANDRA-3579" title="AssertionError in
hintedhandoff - 1.0.5" class="issue-link"
data-issue-key="CASSANDRA-3579"><del>CASSANDRA-3579</del></a> 


New Comment: 
<blockquote>RemoveTest: it fails, saying that '/127.0.0.1:7010 is in use by another
process' (consistently). </blockquote>FWIW, I get a timeout instead (on Windows). 


New Comment: 
I get the same thing Sylvain does in RemoveTest : '/127.0.0.1:7010 is in use by another
process' on Mac OS X. 


New Comment: 
Same 'in use by another process' under linux.  There is definitely no other process. 


New Comment: 
RemoveTest is trying to stop/start MessagingService multiple times in the same suite (see
setup/tearDown methods).  My guess is that worked well enough pre-<a
href="https://issues.apache.org/jira/browse/CASSANDRA-3335" title="ThreadPoolExecutor
creates threads as non-daemon and will block on shutdown by default" class="issue-link"
data-issue-key="CASSANDRA-3335"><del>CASSANDRA-3335</del></a>.  I'll see about making it
happy again, although this feels fragile. 


New Comment: 
Pushed 452ddf63c530fc573551e6fc9c79c1a876f11dd0 with the socket teardown code added back. 
Now it's hitting <tt>java.io.IOException: Failed to delete
c:\Users\Jonathan\projects\cassandra\git\build\test\cassandra\commitlog\CommitLog-1326319330071.log</tt>.
 Progress? 


New Comment: 
User error...  it was my paused Cassandra instance in my IDE holding that CL segment open.
 With that figured out, I also pushed 1e750138177e9cd9cbd6537451a4b5cd301dab3a which
allows MS to be restarted by RemoveTest.All the tests now pass for me, except
RecoveryManagerTruncateTest which has failed on Windows for 0.8+. 


New Comment: 
I still had intermittent failure of columnFamilyStoreTest, but because
SystemTable.isIndexBuilt() was suffering from the same 'I forgot to expunge tombstones'
problem than SystemTable.loadTokens(). I took on myself to commit the same fix for that
instance directly (and check no other method had this problem).So closing this as all
tests are now passing.However I'd be interested to know if anyone else is seeing the
'unable to create link' stack trace during tests, because if so we should probably open
another ticket to investigate. 


New Comment: 
Building Cassandra 1.0.7 I get one test failure:<div class="code panel"
style="border-width: 1px;"><div class="codeContent panelContent"><pre
class="code-java"><span class="code-object">Class</span>
org.apache.cassandra.db.compaction.CompactionsTestName	Tests	Errors	Failures	Time(s)	Time
Stamp	HostCompactionsTest	1	1	0	0.000	2012-01-19T12:52:18	mbutler-OptiPlex-990TestsName	Status	Type	Time(s)testSuperColumnCompactions	Error	Timeout
occurred. Please note the time in the report does not reflect the time until the
timeout.junit.framework.AssertionFailedError: Timeout occurred. Please note the time in
the report does not reflect the time until the timeout.	0.001</pre></div></div>and also
see exceptions being thrown saying it was unable to create a hard link<div class="code
panel" style="border-width: 1px;"><div class="codeContent panelContent"><pre
class="code-java">  [junit]  WARN 13:53:33,357 Overriding RING_DELAY to 1000ms    [junit]
ERROR 13:53:39,794 Unable to create hard link    [junit] com.sun.jna.LastErrorException:
errno was 17    [junit] 	at org.apache.cassandra.utils.CLibrary.link(Native Method)   
[junit] 	at org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)   
[junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:39,796 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Standard1-hc-1-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Standard1-hc-1-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Standard1-hc-1-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Standard1-hc-1-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ERROR 13:53:39,797 Unable to create hard link    [junit]
com.sun.jna.LastErrorException: errno was 17    [junit] 	at
org.apache.cassandra.utils.CLibrary.link(Native Method)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:39,798 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Standard1-hc-2-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Standard1-hc-2-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Standard1-hc-2-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Standard1-hc-2-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ERROR 13:53:41,501 Unable to create hard link    [junit]
com.sun.jna.LastErrorException: errno was 17    [junit] 	at
org.apache.cassandra.utils.CLibrary.link(Native Method)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:41,502 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1.626972746864617465-hc-1-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1.626972746864617465-hc-1-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1.626972746864617465-hc-1-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1.626972746864617465-hc-1-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ERROR 13:53:41,504 Unable to create hard link    [junit]
com.sun.jna.LastErrorException: errno was 17    [junit] 	at
org.apache.cassandra.utils.CLibrary.link(Native Method)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:41,505 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1-hc-1-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1-hc-1-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1-hc-1-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1-hc-1-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ERROR 13:53:41,802 Unable to create hard link    [junit]
com.sun.jna.LastErrorException: errno was 17    [junit] 	at
org.apache.cassandra.utils.CLibrary.link(Native Method)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:41,802 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1.626972746864617465-hc-2-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1.626972746864617465-hc-2-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1.626972746864617465-hc-2-Digest.sha1 to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1.626972746864617465-hc-2-Digest.sha1
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ERROR 13:53:41,804 Unable to create hard link    [junit]
com.sun.jna.LastErrorException: errno was 17    [junit] 	at
org.apache.cassandra.utils.CLibrary.link(Native Method)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:146)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)   
[junit] 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit]
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] ERROR 13:53:41,805 Fatal exception
in thread <span class="code-object">Thread</span>[NonPeriodicTasks:1,5,main]    [junit]
java.lang.RuntimeException: java.io.IOException: Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1-hc-1-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1-hc-1-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.FBUtilities.unchecked(FBUtilities.java:689)    [junit] 	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:34)    [junit] 	at
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)    [junit] 	at
java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)    [junit] 	at
java.util.concurrent.FutureTask.run(FutureTask.java:138)    [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
   [junit] 	at
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
   [junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)   
[junit] 	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)    [junit]
	at java.lang.<span class="code-object">Thread</span>.run(<span
class="code-object">Thread</span>.java:662)    [junit] Caused by: java.io.IOException:
Unable to create hard link from
build/test/cassandra/data/Keyspace1/Indexed1-hc-1-Filter.db to
/home/mbutler/workspace/vscc/vscc-oss-components/cassandra-1.0.7/build/test/cassandra/data/Keyspace1/backups/Indexed1-hc-1-Filter.db
(errno 17)    [junit] 	at
org.apache.cassandra.utils.CLibrary.createHardLink(CLibrary.java:160)    [junit] 	at
org.apache.cassandra.io.sstable.SSTableReader.createLinks(SSTableReader.java:833)   
[junit] 	at org.apache.cassandra.db.DataTracker$1.runMayThrow(DataTracker.java:161)   
[junit] 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)   
[junit] 	... 8 more    [junit] ------------- ----------------
---------------</pre></div></div>Any suggestions on a fix? 


